### PR TITLE
Analyzer: Add support for Yarn 2+

### PR DESCRIPTION
**Pull request by nnobelis**
Yarn2 is a complete rewrite of Yarn: amongst other changes,
'node_modules' file hierarchy is not used anymore and the Yarn2+
executable is delivered alongside the project using it in the
`.yarn/releases` subdirectory.

To be able to use Yarn1 and Yarn2 on the same system, it is recommended
to install Yarn1 globally, and deliver Yarn2+ alongside the project
sources (see [1]). This means that the Yarn2+ executable won't be
available in ORT's Docker image. Therefore, this Yarn2 Package Manager
find the Yarn2+ executable present in the scanned project code
repository and runs it.

To be able to reuse all the work that has been done in the Npm package
manager to map `package.json` files to ORT's packages, the Yarn2 package
manager uses an option called `nodeLinker` to generate the
'node_modules' file hierarchy. Then it delegates the mapping to the Npm
package manager.

[1]: https://stackoverflow.com/a/59934413/5321324

See also: https://github.com/oss-review-toolkit/ort/issues/2283.

### CODE REVIEW

File: analyzer/src/funTest/assets/projects/synthetic/yarn2-workspaces-expected-output.yml [hunk 1]
```diff
@@ -0,0 +1,175 @@
+---
```
Comment (MarcelBochtler) [inline analyzer/src/funTest/assets/projects/synthetic/yarn2-workspaces-expected-output.yml:1]: Commit message:
> is delivered alongside the project using it in the `.yarn/releases` subdirectory.

This sounds grammatically wrong.

Comment (MarcelBochtler) [inline analyzer/src/funTest/assets/projects/synthetic/yarn2-workspaces-expected-output.yml:1]: Commit message nit:
> Therefore, this Yarn2 Package Manager find the Yarn2+ executable present in the scanned project code repository and runs it.

Add missing `s` to find.

File: analyzer/src/main/kotlin/managers/Yarn2.kt [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+        override fun create(
+            analysisRoot: File,
+            analyzerConfig: AnalyzerConfiguration,
+            repoConfig: RepositoryConfiguration
+        ) = Yarn2(managerName, analysisRoot, analyzerConfig, repoConfig)
+    }
+
+    /**
+     * The Yarn2+ executable is not available in ORT's Docker image: we use the program shipped in .yarn/releases.
```
Comment (MarcelBochtler) [inline analyzer/src/main/kotlin/managers/Yarn2.kt:57]: As ORT is not only run in Docker, I'd rephrase this to something that just says that the executable within the project is reused.

File: analyzer/src/main/kotlin/managers/Yarn2.kt [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+     * The resource file `.yarnrc.yml` contains the path to the executable under the property name `yarnPath` defined
+     * for current project e.g. yarnPath: .yarn/releases/yarn-3.2.1.cjs.
```
Comment (MarcelBochtler) [inline analyzer/src/main/kotlin/managers/Yarn2.kt:59]: for _the_ current project

File: analyzer/src/main/kotlin/managers/Yarn2.kt [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+    /**
+     * A NPM package manager, used to call NPM when Yarn2+ does not provide enough information.
+     */
+    private val delegateNpmAnalyzer = Npm("npm", analysisRoot, analyzerConfig, repoConfig)
+
+    companion object {
+        const val YARN2_RESOURCE_FILE = ".yarnrc.yml"
+
+        // The property in `.yarnrc.yml`containing the path to the Yarn 2+ executable.
```
Comment (MarcelBochtler) [inline analyzer/src/main/kotlin/managers/Yarn2.kt:72]: > `Yarn 2+`

Please be consistent in using a space before the version.

**Review by MarcelBochtler — CHANGES_REQUESTED** [review]

File: analyzer/src/main/kotlin/managers/Yarn2.kt [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+        return workingDir?.let {
+            yarn2CommandsByPath.getOrPut(it) {
+                val yarnConfig = yamlMapper.readTree(workingDir.resolve(YARN2_RESOURCE_FILE))
+                val yarnCommand = yarnConfig.get(YARN_PATH_PROPERTY_NAME)
+                yarnCommand?.let { command ->
+                    val yarnExecutable = workingDir.resolve(command.textValue())
+                    // TODO: Yarn2 executable is a `cjs` file. Check if under Windows it needs to be run with `node`.
+
+                    // TODO: This is a security risk to blindly run code coming from a repository other than ORT's. ORT
+                    //  should download the Yarn2 binary from the official repository and run it.
```
Comment (MarcelBochtler) [inline analyzer/src/main/kotlin/managers/Yarn2.kt:88]: Just an idea: Maybe it is possible to check for the hash to ensure that the executable is valid?

File: analyzer/src/main/kotlin/managers/Yarn2.kt [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+
+    override fun hasLockFile(projectDir: File) = hasYarnLockFile(projectDir)
+
+    override fun command(workingDir: File?): String {
+        return workingDir?.let {
+            yarn2CommandsByPath.getOrPut(it) {
+                val yarnConfig = yamlMapper.readTree(workingDir.resolve(YARN2_RESOURCE_FILE))
+                val yarnCommand = yarnConfig.get(YARN_PATH_PROPERTY_NAME)
+                yarnCommand?.let { command ->
+                    val yarnExecutable = workingDir.resolve(command.textValue())
```
Comment (MarcelBochtler) [inline analyzer/src/main/kotlin/managers/Yarn2.kt:84]: On Unix systems you might need to check if the `yarnExecutable` has the executable bit set.

File: analyzer/src/main/kotlin/managers/Yarn2.kt [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+        definitionFiles.forEach { checkVersion(it.parentFile) }
+
+    override fun runInstall(workingDir: File): ProcessCapture =
+        // It should be enough for Yarn2+ to parse the lockfile to get the dependencies, and, if the former doesn't
+        // exist, run `yarn install --mode update-lockfile` to generate it (since yarn version 3.0.0-rc.10).
+        // Unfortunately, the lockfile doesn't contain the information required to build ORT's packages. An alternative
+        // could be to run `npm show --json` on each package present in the lockfile and rebuild a node_modules
+        // file hierarchy to be able to reuse the code present in the Npm package manager.
+        // Fortunately, an easier solution exist: by setting the option `nodeLinker` in the `yarnrc` file or as an
+        // environment variable, one can instruct Yarn to generate the node_modules files hierarchy.
```
Comment (MarcelBochtler) [inline analyzer/src/main/kotlin/managers/Yarn2.kt:117]: As the alternatives are not used in this code, how about removing this part of the comment?

File: analyzer/src/main/kotlin/managers/Yarn2.kt [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+        run("install", workingDir = workingDir, environment = mapOf("YARN_NODE_LINKER" to "node-modules"))
+
+    override fun getRemotePackageDetails(workingDir: File, packageName: String): JsonNode {
+        // Yarn2+ 'info' command does not provide enough information so the default behaviour of 'npm view --json' is
```
Comment (MarcelBochtler) [inline analyzer/src/main/kotlin/managers/Yarn2.kt:121]: `behaviour` -> `behavior`

**Review by MarcelBochtler — CHANGES_REQUESTED** [review]

File: analyzer/src/main/kotlin/managers/Yarn2.kt (hunk 7 above)
Comment (nnobelis) [reply analyzer/src/main/kotlin/managers/Yarn2.kt:117]: I wanted to explain why there is this env variable. Can we keep it ?

File: analyzer/src/main/kotlin/managers/Yarn2.kt (hunk 5 above)
Comment (nnobelis) [reply analyzer/src/main/kotlin/managers/Yarn2.kt:88]: This is a gook idea but I don't know where to find the reference hash.

They don't put the hashes on their Github:
https://github.com/yarnpkg/berry/releases

They have no hash nor index page on the repo e.g. 
https://repo.yarnpkg.com/3.1.0/packages/yarnpkg-cli/bin/yarn.js

File: cli/src/main/resources/log4j2.xml [hunk 9]
```diff
@@ -23,5 +23,6 @@
     <Logger name="org.ossreviewtoolkit.scanner.scanners.fossid.FossIdConfig" level="info"/>
     <Logger name="org.ossreviewtoolkit.clients.fossid.FossIdRestService" level="info"/>
     <Logger name="org.ossreviewtoolkit.reporter.reporters.fossid.FossIdReporter" level="info"/>
+    <Logger name="org.ossreviewtoolkit.analyzer.managers.Yarn2" level="debug"/>
```
Comment (MarcelBochtler) [inline cli/src/main/resources/log4j2.xml:26]: Why is this required?

Comment (nnobelis) [reply cli/src/main/resources/log4j2.xml]: Removed.

**Comment by MarcelBochtler** [comment]
Please also update the README.md:
https://github.com/oss-review-toolkit/ort/blob/bdf48a910ab515d9e05807cd758838aa7cd64488/README.md?plain=1#L342-L343

File: analyzer/src/funTest/kotlin/managers/Yarn2FunTest.kt [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+
+                val result = resolveDependencies(projectDir)
+
+                val expectedResult = getExpectedResult(projectDir, "yarn2-expected-output.yml")
+                result shouldBe expectedResult
+            }
+
+            "resolve workspace dependencies correctly" {
+                // This test case illustrates the lack of Yarn workspaces support, in particular not all workspace
+                // dependencies get assigned to a scope.
```
Comment (sschuberth) [inline analyzer/src/funTest/kotlin/managers/Yarn2FunTest.kt:71]: What does this mean exactly? Does it mean that the expected result is actually wrong, i.e. it does not reflect reality?

File: analyzer/src/main/kotlin/managers/Yarn2.kt [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+
+        override fun create(
+            analysisRoot: File,
+            analyzerConfig: AnalyzerConfiguration,
+            repoConfig: RepositoryConfiguration
+        ) = Yarn2(managerName, analysisRoot, analyzerConfig, repoConfig)
+    }
+
+    /**
+     * The Yarn2+ is not installed globally: we use the program shipped by the project in .yarn/releases. The resource
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/Yarn2.kt:58]: - "The Yarn 2+ *executable* ..."
- Start capitalized after a colon. But also avoid "we" and use passive voice instead.
- Put `.yarn/releases` in backticks.

File: analyzer/src/main/kotlin/managers/Yarn2.kt [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+     * file `.yarnrc.yml` contains the path to the executable under the property name `yarnPath` defined
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/Yarn2.kt:59]: "under the property name" sounds wrong. Better something like "The value of the `yarnPath` property defines the path ...".

File: analyzer/src/main/kotlin/managers/Yarn2.kt [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+     * for the current project e.g. yarnPath: .yarn/releases/yarn-3.2.1.cjs.
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/Yarn2.kt:60]: Put `yarnPath: .yarn/releases/yarn-3.2.1.cjs` in backticks.

File: analyzer/src/main/kotlin/managers/Yarn2.kt [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+     * This map holds the mapping between the directory and their Yarn2 executables.
+     */
+    private val yarn2CommandsByPath: MutableMap<File, String> = mutableMapOf()
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/Yarn2.kt:63]: `yarn2ExecutablesByPath`

File: analyzer/src/main/kotlin/managers/Yarn2.kt [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+
+    /**
+     * A NPM package manager, used to call NPM when Yarn2+ does not provide enough information.
+     */
+    private val delegateNpmAnalyzer = Npm("npm", analysisRoot, analyzerConfig, repoConfig)
+
+    companion object {
+        const val YARN2_RESOURCE_FILE = ".yarnrc.yml"
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/Yarn2.kt:71]: Please add a comment here, too. Or drop the one in line 73. But be consistent.

File: analyzer/src/main/kotlin/managers/Yarn2.kt [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+
+        // The property in `.yarnrc.yml`containing the path to the Yarn2+ executable.
+        const val YARN_PATH_PROPERTY_NAME = "yarnPath"
+    }
+
+    override fun hasLockFile(projectDir: File) = hasYarnLockFile(projectDir)
+
+    override fun command(workingDir: File?): String {
+        return workingDir?.let {
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/Yarn2.kt:80]: Maybe return early `if (workingDir == null)` to reduce indentation here.

File: analyzer/src/main/kotlin/managers/Yarn2.kt [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+                val yarnConfig = yamlMapper.readTree(workingDir.resolve(YARN2_RESOURCE_FILE))
+                val yarnCommand = yarnConfig.get(YARN_PATH_PROPERTY_NAME)
+                yarnCommand?.let { command ->
+                    val yarnExecutable = workingDir.resolve(command.textValue())
+
+                    // TODO: Yarn2 executable is a `cjs` file. Check if under Windows it needs to be run with `node`.
+
+                    // TODO: This is a security risk to blindly run code coming from a repository other than ORT's. ORT
+                    //  should download the Yarn2 binary from the official repository and run it.
+                    require(yarnExecutable.exists()) {
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/Yarn2.kt:91]: Use `isFile` to actually check for a file.

File: analyzer/src/main/kotlin/managers/Yarn2.kt [hunk 18]
```diff
@@ -0,0 +1,141 @@
...
+                        "The Yarn2+ program '${yarnExecutable.name}' does not exist."
+                    }
+
+                    if (!yarnExecutable.canExecute()) {
+                        log.warn {
+                            "The Yarn2+ program '${yarnExecutable.name}' should be executable. Changing its rights."
+                        }
+                        require(yarnExecutable.setExecutable(true)) {
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/Yarn2.kt:99]: Nit: Blank line before this one.

File: analyzer/src/main/kotlin/managers/Yarn2.kt [hunk 19]
```diff
@@ -0,0 +1,141 @@
...
+                            "Cannot set the Yarn2+ program to be executable."
+                        }
+                    }
+                    yarnExecutable.absolutePath
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/Yarn2.kt:103]: Nit: Blank line before this one.

File: analyzer/src/main/kotlin/managers/Yarn2.kt [hunk 20]
```diff
@@ -0,0 +1,141 @@
...
+                } ?: throw IllegalStateException("No Yarn2+ executable could be found in 'yarnrc.yml'.")
+            }
+        } ?: "yarn2+ executable' in .yarn/releases/ and configured in .yarnrc.yml"
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/Yarn2.kt:106]: This is an awkward string to return here, as it's not a command to execute.

File: analyzer/src/main/kotlin/managers/Yarn2.kt [hunk 21]
```diff
@@ -0,0 +1,141 @@
...
+    }
+
+    override fun getVersion(workingDir: File?): String =
+        // `getVersion` with a `null` parameter is called once by the Analyzer to get the version of the global tool.
+        // For Yarn2+, the version is specific to each definition file being scanned therefore a global version doesn't
+        // apply.
+        // TODO: An alternative would be to collate the versions of all tools in `yarn2CommandsByPath`.
+        if (workingDir == null) "N/A" else super.getVersion(workingDir)
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/Yarn2.kt:114]: The convention is to return an empty string if the version is unknown / cannot be determined.

File: analyzer/src/main/kotlin/managers/Yarn2.kt [hunk 22]
```diff
@@ -0,0 +1,141 @@
...
+        // file hierarchy to be able to reuse the code present in the Npm package manager.
+        // Fortunately, an easier solution exist: by setting the option `nodeLinker` in the `yarnrc` file or as an
+        // environment variable, one can instruct Yarn to generate the node_modules files hierarchy.
+        run("install", workingDir = workingDir, environment = mapOf("YARN_NODE_LINKER" to "node-modules"))
+
+    override fun getRemotePackageDetails(workingDir: File, packageName: String): JsonNode {
+        // Yarn2+ 'info' command does not provide enough information so the default behavior of 'npm view --json' is
+        // kept. Note that it is mandatory to override this function as the super function calls `run()` which would
+        // call Yarn.
+        val process = delegateNpmAnalyzer.run(workingDir, "view", "--json", packageName)
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/Yarn2.kt:138]: Wouldn't it make more sense to delegate to Yarn 1 and `yarn info --json` here?

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: analyzer/src/funTest/kotlin/managers/Yarn2FunTest.kt [hunk 23]
```diff
@@ -0,0 +1,84 @@
...
+            path = vcsPath,
+            custom = mapOf("<REPLACE_RAW_URL>" to vcsUrl)
+        )
+    }
+
+    private fun resolveDependencies(projectDir: File): String {
+        val packageFile = projectDir.resolve("package.json")
+        val result = createYarn2().resolveSingleProject(packageFile, resolveScopes = true)
+        return result.toYaml()
+    }
```
[the remark this answers is not in the document]
Comment (nnobelis) [reply analyzer/src/funTest/kotlin/managers/Yarn2FunTest.kt:56]: Done. But this code is strongly inspired by `YarnFunTest`.

File: analyzer/src/funTest/kotlin/managers/Yarn2FunTest.kt (hunk 10 above)
Comment (nnobelis) [reply analyzer/src/funTest/kotlin/managers/Yarn2FunTest.kt:71]: This is a test that I (dumbly) copied from `YarnFunTest`. I think it should be rewritten because Yarn 2 _does_ support workspaces.
What do you think ? Should I remove the whole test for now and put a TODO ?

File: analyzer/src/main/kotlin/managers/Yarn2.kt (hunk 20 above)
Comment (nnobelis) [reply analyzer/src/main/kotlin/managers/Yarn2.kt:106]: This is because `workingDir` is `null` only when called by `ort requirements`.

Therefore this output is made for the output of `ort requirements`

```
	- Stack: Requires 'stack' in version >=2.1.1. Tool not found.
	- Yarn: Requires 'yarn' in version >=1.3.0 <1.23.0. Tool not found.
	- Yarn2: Requires 'yarn2+ executable' in .yarn/releases/ and configured in .yarnrc.yml' in version >=2.0.0. Tool not found.
```

Since this is an ORT quirk, I am open for opinion here.

Comment (MarcelBochtler) [reply analyzer/src/main/kotlin/managers/Yarn2.kt]: IMO, Yarn2 should not be mentioned at all in the `requirements` command, as no installation is required.
`Gradle` is also not mentioned in `requirements`. Then this String could also be removed.

I think it's dangerous to return this String as a command. I think you cannot rely on `command` without a `workingDir` will only be called from `requirements`.

File: analyzer/src/main/kotlin/managers/Yarn2.kt (hunk 22 above)
Comment (nnobelis) [reply analyzer/src/main/kotlin/managers/Yarn2.kt:142]: > Getting started with Yarn 2 requires you to have an up to date version of Node.js, at minimum Node.js 10.

From [1].
And when you have a Node installation, you have automatically NPM.
Therefore you always have NPM on an Yarn2 project. Right ?

[1]: https://snyk.io/blog/yarn-2-intro/

**Review by MarcelBochtler — APPROVED** [review]

**Comment by nnobelis** [comment]
@MarcelBochtler since the test failure is unrelated (dotnet), can we merge this ?

**Comment by sschuberth** [comment]
Merging despite the unrelated `DotNetFunTest` failure.

**Comment by nnobelis** [comment]
@sschuberth  can I close https://github.com/oss-review-toolkit/ort/issues/2283 ?

**Comment by sschuberth** [comment]
> @sschuberth can I close #2283 ?

Ah, crap, that should have gone to the commit message as an [auto-close link](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue). But yes, you can close it manually now with a reference to this PR.
